[clr-interp] Stackoverflow on x64

### Description

This happens in our test suite:

```
Stack overflow.
Repeated 7349 times:
--------------------------------
   at System.Runtime.InteropServices.DynamicInterfaceCastableHelpers.GetInterfaceImplementation(System.Runtime.InteropServices.IDynamicInterfaceCastable, System.RuntimeType)
--------------------------------
   at Xamarin.Tests.AvailabilityExtensions.IsAvailable(System.Collections.Generic.IEnumerable`1<System.ValueTuple`3<System.Runtime.Versioning.OSPlatformAttribute,Xamarin.Utils.ApplePlatform,System.Version>>, Xamarin.Tests.PlatformInfo, Xamarin.Utils.ApplePlatform)
   at Xamarin.Tests.AvailabilityExtensions.IsAvailable(System.Collections.Generic.IEnumerable`1<System.Runtime.Versioning.OSPlatformAttribute>, Xamarin.Tests.PlatformInfo)
   at Xamarin.Tests.AvailabilityExtensions.IsAvailable(System.Reflection.ICustomAttributeProvider, Xamarin.Tests.PlatformInfo)
   at Xamarin.Tests.AvailabilityExtensions.IsAvailableOnHostPlatform(System.Reflection.ICustomAttributeProvider)
   at Introspection.ApiBaseTest.SkipDueToAttribute(System.Reflection.MemberInfo)
   at Introspection.ApiCMAttachmentTest.Skip(System.Type)
[...]
```

### Reproduction Steps

Needs Xcode 26.1.1 installed in /Applications/Xcode_26.1.1.app.

```shell
$ git clone --recursive https://github.com/dotnet/macios
$ cd macios
$ git checkout dev/rolf/net11.0-coreclr
$ git reset --hard 0129a5d735b6ff4947c354e7ffa6233c6fa4a2c9
$ ./configure --disable-all-platforms --enable-maccatalyst
$ make all -j8
$ make install -j8
$ make build run-bare TEST_VARIATION=coreclr RID=maccatalyst-x64 -C tests/introspection/dotnet/MacCatalyst
```



### Expected behavior

Successful test execution.

### Actual behavior

<summary>Stack overflow.

<details>

```
$ make build run-bare TEST_VARIATION=coreclr RID=maccatalyst-x64
Found ccache on the system, enabling it
Detected the macios-adr repository, automatically enabled the Xamarin build
2025-11-14 16:16:43.642 introspection[86919:4310425] Microsoft.MacCatalyst: Invalid IDE Port: -1
2025-11-14 16:16:47.496 introspection[86919:4309880] Queuing test loading on background thread
2025-11-14 16:16:47.503 introspection[86919:4310554] Loading tests on background thread
2025-11-14 16:16:47.580 introspection[86919:4310554] Loaded tests on background thread, queuing UI update on main thead
2025-11-14 16:16:47.907 introspection[86919:4309880] Updating UI on main thead
2025-11-14 16:16:48.089 introspection[86919:4309880] AutoRun ()
2025-11-14 16:16:48.089 introspection[86919:4309880] AutoRun (): queuing test run on main thread
2025-11-14 16:16:48.095 introspection[86919:4309880] AutoRun (): running tests on main thread
2025-11-14 16:16:48.129 introspection[86919:4309880] [Runner executing:	Run Everything]
2025-11-14 16:16:48.129 introspection[86919:4309880] [MonoTouch Version:	26.0.11017]
2025-11-14 16:16:48.130 introspection[86919:4309880] [Assembly:	Microsoft.MacCatalyst.dll (64 bits)]
2025-11-14 16:16:48.130 introspection[86919:4309880] [GC:	sgen]
2025-11-14 16:16:48.130 introspection[86919:4309880] [iPad:	iPadOS v18.7]
2025-11-14 16:16:48.130 introspection[86919:4309880] [Device Name:	iPad]
2025-11-14 16:16:48.131 introspection[86919:4309880] [Device UDID:	unknown]
2025-11-14 16:16:48.158 introspection[86919:4309880] [Device Locale:	en_US@rg=eszzzz]
2025-11-14 16:16:48.160 introspection[86919:4309880] [Device Date/Time:	11/14/2025 4:16:48 PM]
2025-11-14 16:16:48.161 introspection[86919:4309880] [Bundle:	com.xamarin.introspection]
2025-11-14 16:16:48.171 introspection[86919:4310580] introspection
2025-11-14 16:16:48.175 introspection[86919:4310580] Introspection
2025-11-14 16:16:48.175 introspection[86919:4310580] Introspection.ApiCMAttachmentTest
2025-11-14 16:16:48.298 introspection[86919:4310580] 	[PASS] CheckAttachments
Stack overflow.
Repeated 59038 times:
--------------------------------
   at System.Runtime.InteropServices.DynamicInterfaceCastableHelpers.GetInterfaceImplementation(System.Runtime.InteropServices.IDynamicInterfaceCastable, System.RuntimeType)
--------------------------------
   at Xamarin.Tests.AvailabilityExtensions.IsAvailable(System.Collections.Generic.IEnumerable`1<System.ValueTuple`3<System.Runtime.Versioning.OSPlatformAttribute,Xamarin.Utils.ApplePlatform,System.Version>>, Xamarin.Tests.PlatformInfo, Xamarin.Utils.ApplePlatform)
   at Xamarin.Tests.AvailabilityExtensions.IsAvailable(System.Collections.Generic.IEnumerable`1<System.Runtime.Versioning.OSPlatformAttribute>, Xamarin.Tests.PlatformInfo)
   at Xamarin.Tests.AvailabilityExtensions.IsAvailable(System.Reflection.ICustomAttributeProvider, Xamarin.Tests.PlatformInfo)
   at Xamarin.Tests.AvailabilityExtensions.IsAvailableOnHostPlatform(System.Reflection.ICustomAttributeProvider)
   at Introspection.ApiBaseTest.SkipDueToAttribute(System.Reflection.MemberInfo)
   at Introspection.ApiCMAttachmentTest.Skip(System.Type)
   at Introspection.ApiCMAttachmentTest.CheckFailAttachments()
   at System.RuntimeMethodHandle.InvokeMethod(System.Object, Void**, System.Signature, Boolean)
   at System.Reflection.MethodBaseInvoker.InterpretedInvoke_Method(System.Object, IntPtr*)
   at System.Reflection.MethodBaseInvoker.InvokeWithNoArgs(System.Object, System.Reflection.BindingFlags)
   at System.Reflection.RuntimeMethodInfo.Invoke(System.Object, System.Reflection.BindingFlags, System.Reflection.Binder, System.Object[], System.Globalization.CultureInfo)
   at System.Reflection.MethodBase.Invoke(System.Object, System.Object[])
   at NUnit.Framework.Internal.Reflect.InvokeMethod(System.Reflection.MethodInfo, System.Object, System.Object[])
   at NUnit.Framework.Internal.MethodWrapper.Invoke(System.Object, System.Object[])
   at NUnit.Framework.Internal.Commands.TestMethodCommand.InvokeTestMethod(NUnit.Framework.Internal.TestExecutionContext)
   at NUnit.Framework.Internal.Commands.TestMethodCommand.RunTestMethod(NUnit.Framework.Internal.TestExecutionContext)
   at NUnit.Framework.Internal.Commands.TestMethodCommand.Execute(NUnit.Framework.Internal.TestExecutionContext)
   at NUnit.Framework.Internal.Execution.SimpleWorkItem.PerformWork()
   at NUnit.Framework.Internal.Execution.WorkItem.RunOnCurrentThread()
   at NUnit.Framework.Internal.Execution.WorkItem.Execute()
   at NUnit.Framework.Internal.Execution.MainThreadWorkItemDispatcher.Dispatch(NUnit.Framework.Internal.Execution.WorkItem)
   at NUnit.Framework.Internal.Execution.CompositeWorkItem.RunChildren()
   at NUnit.Framework.Internal.Execution.CompositeWorkItem.PerformWork()
   at NUnit.Framework.Internal.Execution.WorkItem.RunOnCurrentThread()
   at NUnit.Framework.Internal.Execution.WorkItem.Execute()
   at NUnit.Framework.Internal.Execution.MainThreadWorkItemDispatcher.Dispatch(NUnit.Framework.Internal.Execution.WorkItem)
   at NUnit.Framework.Internal.Execution.CompositeWorkItem.RunChildren()
   at NUnit.Framework.Internal.Execution.CompositeWorkItem.PerformWork()
   at NUnit.Framework.Internal.Execution.WorkItem.RunOnCurrentThread()
   at NUnit.Framework.Internal.Execution.WorkItem.Execute()
   at NUnit.Framework.Internal.Execution.MainThreadWorkItemDispatcher.Dispatch(NUnit.Framework.Internal.Execution.WorkItem)
   at NUnit.Framework.Internal.Execution.CompositeWorkItem.RunChildren()
   at NUnit.Framework.Internal.Execution.CompositeWorkItem.PerformWork()
   at NUnit.Framework.Internal.Execution.WorkItem.RunOnCurrentThread()
   at NUnit.Framework.Internal.Execution.WorkItem.Execute()
   at NUnit.Framework.Internal.Execution.MainThreadWorkItemDispatcher.Dispatch(NUnit.Framework.Internal.Execution.WorkItem)
   at NUnit.Framework.Internal.Execution.MainThreadWorkItemDispatcher.Start(NUnit.Framework.Internal.Execution.WorkItem)
   at NUnit.Framework.Api.NUnitTestAssemblyRunner.StartRun(NUnit.Framework.Interfaces.ITestListener)
   at NUnit.Framework.Api.NUnitTestAssemblyRunner+<>c__DisplayClass38_0.<RunAsync>b__0()
   at NUnit.Framework.Api.NUnitTestAssemblyRunner.WrapInNUnitCallContext(System.Action)
   at NUnit.Framework.Api.NUnitTestAssemblyRunner.RunAsync(NUnit.Framework.Interfaces.ITestListener, NUnit.Framework.Interfaces.ITestFilter)
   at NUnit.Framework.Api.NUnitTestAssemblyRunner.Run(NUnit.Framework.Interfaces.ITestListener, NUnit.Framework.Interfaces.ITestFilter)
   at MonoTouch.NUnit.UI.BaseTouchRunner.Run(NUnit.Framework.Internal.Test)
   at MonoTouch.NUnit.UI.BaseTouchRunner.Run()
   at MonoTouch.NUnit.UI.BaseTouchRunner.<AutoRun>b__49_0()
   at MonoTouch.NUnit.UI.TouchRunner+<>c__DisplayClass19_0.<ExecuteOnMainThread>b__0()
   at Foundation.NSAsyncActionDispatcher.Apply()
   at System.RuntimeMethodHandle.InvokeMethod(System.Object, Void**, System.Signature, Boolean)
   at System.Reflection.MethodBaseInvoker.InterpretedInvoke_Method(System.Object, IntPtr*)
   at System.Reflection.MethodBaseInvoker.InvokeWithNoArgs(System.Object, System.Reflection.BindingFlags)
   at System.Reflection.RuntimeMethodInfo.Invoke(System.Object, System.Reflection.BindingFlags, System.Reflection.Binder, System.Object[], System.Globalization.CultureInfo)
   at ObjCRuntime.Runtime.InvokeMethod(System.Reflection.MethodBase, System.Object, IntPtr)
   at ObjCRuntime.Runtime.InvokeMethod(MonoObject*, MonoObject*, IntPtr)
   at ObjCRuntime.Runtime.bridge_runtime_invoke_method(MonoObject*, MonoObject*, IntPtr, IntPtr*)
   at UIKit.UIApplication.UIApplicationMain(Int32, System.String[], IntPtr, IntPtr)
   at UIKit.UIApplication.Main(System.String[], System.Type, System.Type)
   at MainClass.Main(System.String[])
make: *** [run-bare] Abort trap: 6
```
</details>
</summary>

### Regression?

No

### Known Workarounds

_No response_

### Configuration

_No response_

### Other information

_No response_

I'm not sure if this is the same bug or not, but our other test suite also crashes, with something that looks like stack overflows: https://gist.github.com/rolfbjarne/fc397ffd355ecfc7f6c89fa48a4e9924

Repro is the same, except the last line, which is:

```shell
$ make build run-bare TEST_VARIATION=coreclr RID=maccatalyst-x64 -C tests/monotouch-test/dotnet/MacCatalyst
```

<summary>Output

<details>

```

2025-11-15 00:12:44.664 monotouchtest[35211:5314588] Microsoft.MacCatalyst: Invalid IDE Port: -1
2025-11-15 00:12:51.846 monotouchtest[35211:5313856] ObjCRuntime.RuntimeException: Detected a protocol (MonoTouchFixtures.JavascriptCore.IMyJavaExporter) inheriting from the JSExport protocol while using the dynamic registrar. It is not possible to export protocols to JavaScriptCore dynamically; the static registrar must be used (add '--registrar:static' to the additional mtouch arguments in the project's iOS Build options to select the static registrar).
2025-11-15 00:13:29.110 monotouchtest[35211:5313856] Queuing test loading on background thread
2025-11-15 00:13:29.115 monotouchtest[35211:5315974] Loading tests on background thread
2025-11-15 00:13:33.182 monotouchtest[35211:5315974] Loaded tests on background thread, queuing UI update on main thead
2025-11-15 00:13:33.183 monotouchtest[35211:5313856] Updating UI on main thead
2025-11-15 00:13:34.070 monotouchtest[35211:5313856] AutoRun ()
2025-11-15 00:13:34.070 monotouchtest[35211:5313856] AutoRun (): queuing test run on main thread
2025-11-15 00:13:34.076 monotouchtest[35211:5313856] AutoRun (): running tests on main thread
2025-11-15 00:13:34.077 monotouchtest[35211:5313856] [Runner executing:	Run Everything]
2025-11-15 00:13:34.077 monotouchtest[35211:5313856] [MonoTouch Version:	26.0.11017]
2025-11-15 00:13:34.078 monotouchtest[35211:5313856] [Assembly:	Microsoft.MacCatalyst.dll (64 bits)]
2025-11-15 00:13:34.078 monotouchtest[35211:5313856] [GC:	sgen]
2025-11-15 00:13:34.078 monotouchtest[35211:5313856] [iPad:	iPadOS v18.7]
2025-11-15 00:13:34.078 monotouchtest[35211:5313856] [Device Name:	iPad]
2025-11-15 00:13:34.078 monotouchtest[35211:5313856] [Device UDID:	unknown]
2025-11-15 00:13:34.094 monotouchtest[35211:5313856] [Device Locale:	en_US@rg=eszzzz]
2025-11-15 00:13:34.096 monotouchtest[35211:5313856] [Device Date/Time:	11/15/2025 12:13:34 AM]
2025-11-15 00:13:34.096 monotouchtest[35211:5313856] [Bundle:	com.xamarin.monotouch-test]
2025-11-15 00:13:59.719 monotouchtest[35211:5317127] monotouchtest
2025-11-15 00:13:59.722 monotouchtest[35211:5317127] LoaderTest
2025-11-15 00:13:59.723 monotouchtest[35211:5317127] LoaderTest : 0.2857 ms
2025-11-15 00:13:59.723 monotouchtest[35211:5317127] MonoTests
2025-11-15 00:13:59.723 monotouchtest[35211:5317127] MonoTests.System
2025-11-15 00:13:59.723 monotouchtest[35211:5317127] MonoTests.System.Net
2025-11-15 00:13:59.723 monotouchtest[35211:5317127] MonoTests.System.Net.Http
2025-11-15 00:13:59.723 monotouchtest[35211:5317127] MonoTests.System.Net.Http.MessageHandlerTest
2025-11-15 00:13:59.725 monotouchtest[35211:5317127] MonoTests.System.Net.Http.MessageHandlerTest.AcceptSslCertificatesServicePointManager
2025-11-15 00:14:04.211 monotouchtest[35211:5317127] 	[PASS] AcceptSslCertificatesServicePointManager(System.Net.Http.HttpClientHandler)
2025-11-15 00:14:06.950 monotouchtest[35211:5317127] 	[PASS] AcceptSslCertificatesServicePointManager(System.Net.Http.NSUrlSessionHandler)
2025-11-15 00:14:06.953 monotouchtest[35211:5317127] MonoTests.System.Net.Http.MessageHandlerTest.AcceptSslCertificatesServicePointManager : 7228.0127 ms
2025-11-15 00:14:06.954 monotouchtest[35211:5317127] MonoTests.System.Net.Http.MessageHandlerTest.AcceptSslCertificatesWithCustomValidationCallbackNSUrlSessionHandler
2025-11-15 00:14:06.958 monotouchtest[35211:5317127] 	[IGNORED] AcceptSslCertificatesWithCustomValidationCallbackNSUrlSessionHandler("https://self-signed.badssl.com/") : OneTimeSetUp: https://github.com/dotnet/macios/issues/21912
2025-11-15 00:14:06.958 monotouchtest[35211:5317127] 	[IGNORED] AcceptSslCertificatesWithCustomValidationCallbackNSUrlSessionHandler("https://wrong.host.badssl.com/") : OneTimeSetUp: https://github.com/dotnet/macios/issues/21912
2025-11-15 00:14:06.958 monotouchtest[35211:5317127] 	[SKIPREASON] https://github.com/dotnet/macios/issues/21912
2025-11-15 00:14:06.959 monotouchtest[35211:5317127] MonoTests.System.Net.Http.MessageHandlerTest.AcceptSslCertificatesWithCustomValidationCallbackNSUrlSessionHandler : 4.6304 ms
2025-11-15 00:14:06.960 monotouchtest[35211:5317127] 	[PASS] AssertDefaultValuesNSUrlSessionHandler
2025-11-15 00:14:06.960 monotouchtest[35211:5317127] MonoTests.System.Net.Http.MessageHandlerTest.DnsFailure
2025-11-15 00:14:07.170 monotouchtest[35211:5317127] 	[PASS] DnsFailure(System.Net.Http.HttpClientHandler)
2025-11-15 00:14:07.288 monotouchtest[35211:5317127] 	[PASS] DnsFailure(System.Net.Http.CFNetworkHandler)
2025-11-15 00:14:07.395 monotouchtest[35211:5317127] 	[PASS] DnsFailure(System.Net.Http.SocketsHttpHandler)
2025-11-15 00:14:07.500 monotouchtest[35211:5317127] 	[PASS] DnsFailure(System.Net.Http.NSUrlSessionHandler)
2025-11-15 00:14:07.500 monotouchtest[35211:5317127] MonoTests.System.Net.Http.MessageHandlerTest.DnsFailure : 540.0092 ms
2025-11-15 00:14:07.501 monotouchtest[35211:5317127] MonoTests.System.Net.Http.MessageHandlerTest.GHIssue16339
2025-11-15 00:14:08.249 monotouchtest[35211:5317127] 	[PASS] GHIssue16339()
2025-11-15 00:14:08.250 monotouchtest[35211:5317127] MonoTests.System.Net.Http.MessageHandlerTest.GHIssue16339 : 748.8319 ms
2025-11-15 00:14:08.250 monotouchtest[35211:5317127] MonoTests.System.Net.Http.MessageHandlerTest.GHIssue8342
2025-11-15 00:14:09.168 monotouchtest[35211:5317127] 	[PASS] GHIssue8342(OK,"mandel","12345678","mandel","12345678")
2025-11-15 00:14:10.287 monotouchtest[35211:5317127] 	[PASS] GHIssue8342(Unauthorized,"mandel","12345678","mandel","87654321")
2025-11-15 00:14:12.929 monotouchtest[35211:5317127] 	[PASS] GHIssue8342(Unauthorized,"mandel","12345678","","")
2025-11-15 00:14:12.930 monotouchtest[35211:5317127] MonoTests.System.Net.Http.MessageHandlerTest.GHIssue8342 : 4680.0683 ms
2025-11-15 00:14:12.931 monotouchtest[35211:5317127] MonoTests.System.Net.Http.MessageHandlerTest.GHIssue8344
2025-11-15 00:14:18.619 monotouchtest[35211:5317127] 	[PASS] GHIssue8344()
2025-11-15 00:14:18.620 monotouchtest[35211:5317127] MonoTests.System.Net.Http.MessageHandlerTest.GHIssue8344 : 5689.525 ms
2025-11-15 00:14:18.621 monotouchtest[35211:5317127] MonoTests.System.Net.Http.MessageHandlerTest.RedirectionWithAuthorizationHeaders
make: *** [run-bare] Segmentation fault: 11
```

</details>
</summary>

--- a/src/coreclr/vm/amd64/asmhelpers.S
+++ b/src/coreclr/vm/amd64/asmhelpers.S
@@ -574,7 +574,7 @@ END_PROLOGUE
         ret
 NESTED_END InterpreterStubRetDoubleI8, _TEXT
 
-NESTED_ENTRY InterpreterStubRetBuff, _TEXT, NoHandler
+NESTED_ENTRY InterpreterStubRetBuffRDI, _TEXT, NoHandler
         push_register   rax // stack alignment
 END_PROLOGUE
         mov             rdi, rax // pTransitionBlock*
@@ -586,7 +586,21 @@ END_PROLOGUE
         mov             rax, qword ptr [rsp + 8 + 8 + __InterpreterStubArgumentRegistersOffset]
         pop_register    rcx
         ret
-NESTED_END InterpreterStubRetBuff, _TEXT
+NESTED_END InterpreterStubRetBuffRDI, _TEXT
+
+NESTED_ENTRY InterpreterStubRetBuffRSI, _TEXT, NoHandler
+        push_register   rax // stack alignment
+END_PROLOGUE
+        mov             rdi, rax // pTransitionBlock*
+        mov             rsi, rbx // the IR bytecode pointer
+        // Load the return buffer address
+        // The 8 + 8 is for the push and the return address slot
+        mov             rdx, qword ptr [rsp + 8 + 8 + __InterpreterStubArgumentRegistersOffset + 8]
+        call            C_FUNC(ExecuteInterpretedMethod)
+        mov             rax, qword ptr [rsp + 8 + 8 + __InterpreterStubArgumentRegistersOffset + 8]
+        pop_register    rcx
+        ret
+NESTED_END InterpreterStubRetBuffRSI, _TEXT
 
 // Routines for passing value type arguments by reference in general purpose registers RDI, RSI, RDX, RCX, R8, R9
 // from native code to the interpreter
@@ -1732,7 +1746,7 @@ END_PROLOGUE
         ret
 NESTED_END CallJittedMethodRetVoid, _TEXT
 
-NESTED_ENTRY CallJittedMethodRetBuff, _TEXT, NoHandler
+NESTED_ENTRY CallJittedMethodRetBuffRDI, _TEXT, NoHandler
         push_nonvol_reg rbp
         mov  rbp, rsp
         alloc_stack 0x10
@@ -1747,7 +1761,24 @@ END_PROLOGUE
         mov rsp, rbp
         pop rbp
         ret
-NESTED_END CallJittedMethodRetBuff, _TEXT
+NESTED_END CallJittedMethodRetBuffRDI, _TEXT
+
+NESTED_ENTRY CallJittedMethodRetBuffRSI, _TEXT, NoHandler
+        push_nonvol_reg rbp
+        mov  rbp, rsp
+        alloc_stack 0x10
+        save_reg_postrsp r10, 0
+END_PROLOGUE
+        sub rsp, rcx // total stack space
+        mov r11, rdi // The routines list
+        mov r10, rsi // interpreter stack args
+        mov rsi, rdx // return buffer
+        call qword ptr [r11]
+        mov r10, [rsp]
+        mov rsp, rbp
+        pop rbp
+        ret
+NESTED_END CallJittedMethodRetBuffRSI, _TEXT
 
 NESTED_ENTRY CallJittedMethodRetDouble, _TEXT, NoHandler
         push_nonvol_reg rbp

--- a/src/coreclr/vm/callstubgenerator.cpp
+++ b/src/coreclr/vm/callstubgenerator.cpp
@@ -1990,7 +1990,7 @@ extern "C" void InterpreterStubRetBuffRSI();
 #else // TARGET_AMD64
 extern "C" void CallJittedMethodRetBuff(PCODE *routines, int8_t*pArgs, int8_t*pRet, int totalStackSize);
 extern "C" void InterpreterStubRetBuff();
-#endif // TARGET_WINDOWS && TARGET_AMD64
+#endif // TARGET_AMD64
 
 #ifdef UNIX_AMD64_ABI
 extern "C" void CallJittedMethodRetI8I8(PCODE *routines, int8_t*pArgs, int8_t*pRet, int totalStackSize);
@@ -2082,7 +2082,7 @@ CallStubHeader::InvokeFunctionPtr CallStubGenerator::GetInvokeFunctionPtr(CallSt
 #else // TARGET_AMD64
         case ReturnTypeBuff:
             INVOKE_FUNCTION_PTR(CallJittedMethodRetBuff);
-#endif // TARGET_WINDOWS && TARGET_AMD64
+#endif // TARGET_AMD64
 #ifdef UNIX_AMD64_ABI
         case ReturnTypeI8I8:
             INVOKE_FUNCTION_PTR(CallJittedMethodRetI8I8);
@@ -2177,7 +2177,7 @@ PCODE CallStubGenerator::GetInterpreterReturnTypeHandler(CallStubGenerator::Retu
 #else // TARGET_AMD64
         case ReturnTypeBuff:
             RETURN_TYPE_HANDLER(InterpreterStubRetBuff);
-#endif // TARGET_WINDOWS && TARGET_AMD64
+#endif // TARGET_AMD64
 #ifdef UNIX_AMD64_ABI
         case ReturnTypeI8I8:
             RETURN_TYPE_HANDLER(InterpreterStubRetI8I8);
@@ -2832,7 +2832,7 @@ CallStubGenerator::ReturnType CallStubGenerator::GetReturnType(ArgIterator *pArg
         }
 #else
         return ReturnTypeBuff;
-#endif
+#endif // TARGET_AMD64
     }
     else
     {

--- a/src/coreclr/vm/callstubgenerator.cpp
+++ b/src/coreclr/vm/callstubgenerator.cpp
@@ -1975,12 +1975,19 @@ extern "C" void InterpreterStubRetVoid();
 extern "C" void InterpreterStubRetDouble();
 extern "C" void InterpreterStubRetI8();
 
-#if defined(TARGET_WINDOWS) && defined(TARGET_AMD64)
+#ifdef TARGET_AMD64
+#ifdef TARGET_WINDOWS
 extern "C" void CallJittedMethodRetBuffRCX(PCODE *routines, int8_t*pArgs, int8_t*pRet, int totalStackSize);
 extern "C" void CallJittedMethodRetBuffRDX(PCODE *routines, int8_t*pArgs, int8_t*pRet, int totalStackSize);
 extern "C" void InterpreterStubRetBuffRCX();
 extern "C" void InterpreterStubRetBuffRDX();
-#else // TARGET_WINDOWS && TARGET_AMD64
+#else // TARGET_WINDOWS
+extern "C" void CallJittedMethodRetBuffRDI(PCODE *routines, int8_t*pArgs, int8_t*pRet, int totalStackSize);
+extern "C" void CallJittedMethodRetBuffRSI(PCODE *routines, int8_t*pArgs, int8_t*pRet, int totalStackSize);
+extern "C" void InterpreterStubRetBuffRDI();
+extern "C" void InterpreterStubRetBuffRSI();
+#endif // TARGET_WINDOWS
+#else // TARGET_AMD64
 extern "C" void CallJittedMethodRetBuff(PCODE *routines, int8_t*pArgs, int8_t*pRet, int totalStackSize);
 extern "C" void InterpreterStubRetBuff();
 #endif // TARGET_WINDOWS && TARGET_AMD64
@@ -2060,12 +2067,19 @@ CallStubHeader::InvokeFunctionPtr CallStubGenerator::GetInvokeFunctionPtr(CallSt
             INVOKE_FUNCTION_PTR(CallJittedMethodRetDouble);
         case ReturnTypeI8:
             INVOKE_FUNCTION_PTR(CallJittedMethodRetI8);
-#if defined(TARGET_WINDOWS) && defined(TARGET_AMD64)
+#ifdef TARGET_AMD64
+#ifdef TARGET_WINDOWS
         case ReturnTypeBuffArg1:
             INVOKE_FUNCTION_PTR(CallJittedMethodRetBuffRCX);
         case ReturnTypeBuffArg2:
             INVOKE_FUNCTION_PTR(CallJittedMethodRetBuffRDX);
-#else // TARGET_WINDOWS && TARGET_AMD64
+#else // TARGET_WINDOWS
+        case ReturnTypeBuffArg1:
+            INVOKE_FUNCTION_PTR(CallJittedMethodRetBuffRDI);
+        case ReturnTypeBuffArg2:
+            INVOKE_FUNCTION_PTR(CallJittedMethodRetBuffRSI);
+#endif // TARGET_WINDOWS
+#else // TARGET_AMD64
         case ReturnTypeBuff:
             INVOKE_FUNCTION_PTR(CallJittedMethodRetBuff);
 #endif // TARGET_WINDOWS && TARGET_AMD64
@@ -2147,12 +2161,20 @@ PCODE CallStubGenerator::GetInterpreterReturnTypeHandler(CallStubGenerator::Retu
             RETURN_TYPE_HANDLER(InterpreterStubRetDouble);
         case ReturnTypeI8:
             RETURN_TYPE_HANDLER(InterpreterStubRetI8);
-#if defined(TARGET_WINDOWS) && defined(TARGET_AMD64)
+#ifdef TARGET_AMD64
         case ReturnTypeBuffArg1:
+#ifdef TARGET_WINDOWS
             RETURN_TYPE_HANDLER(InterpreterStubRetBuffRCX);
+#else
+            RETURN_TYPE_HANDLER(InterpreterStubRetBuffRDI);
+#endif
         case ReturnTypeBuffArg2:
+#ifdef TARGET_WINDOWS
             RETURN_TYPE_HANDLER(InterpreterStubRetBuffRDX);
-#else // TARGET_WINDOWS && TARGET_AMD64
+#else
+            RETURN_TYPE_HANDLER(InterpreterStubRetBuffRSI);
+#endif
+#else // TARGET_AMD64
         case ReturnTypeBuff:
             RETURN_TYPE_HANDLER(InterpreterStubRetBuff);
 #endif // TARGET_WINDOWS && TARGET_AMD64
@@ -2799,7 +2821,7 @@ CallStubGenerator::ReturnType CallStubGenerator::GetReturnType(ArgIterator *pArg
 {
     if (pArgIt->HasRetBuffArg())
     {
-#if defined(TARGET_WINDOWS) && defined(TARGET_AMD64)
+#ifdef TARGET_AMD64
         if (pArgIt->HasThis())
         {
             return ReturnTypeBuffArg2;

--- a/src/coreclr/vm/callstubgenerator.h
+++ b/src/coreclr/vm/callstubgenerator.h
@@ -70,7 +70,7 @@ class CallStubGenerator
         ReturnTypeVoid,
         ReturnTypeI8,
         ReturnTypeDouble,
-#if defined(TARGET_WINDOWS) && defined(TARGET_AMD64)
+#ifdef TARGET_AMD64
         ReturnTypeBuffArg1,
         ReturnTypeBuffArg2,
 #else

